### PR TITLE
bpo-41818: Make test_openpty() avoid unexpected success due to number of rows and/or number of columns being == 0.

### PR DESCRIPTION
**Pull request by 8vasu**
Proposed proper fix for buildbot failures introduced after #22962 merging.

To check if `pty.openpty()` properly sets pty slave size based on the size of current stdin, `PtyTest.test_openpty()` was modifying the size of current stdin first. Let 'x' be the number of rows of stdin; then, we set the number of rows of stdin to some f(x) such that the function f does not have any fixed points; that is, f(x) != x for all x. In #22962, f(x) := x//5; unfortunately, in that case, f does have 0 as a fixed point [ 0//5 = 0 ]. Upon running

```
./python -c "import pty;pty.spawn('./python -m test -v test_pty'.split())"
```

unexpected success was reported on Debian; this indeed was due to the # of (rows, cols) being (0,0). The Gentoo buildbot was probably also running with stdin being a tty of size (0,0). The new choice of f(x) := x+1 solves the problem, since this f has no fixed points [ x+1=x has no solution. ]

TL;DR: instead of letting rows and cols of stdin be the floor of 1/5 of their original values, we now only increment their respective values by 1. Note, however, that the former is a more noticeable change when # of rows, cols are fairly large.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

https://bugs.python.org/issue41818

### CODE REVIEW

**Comment by 8vasu** [comment]
@zware thank you for suggesting that helpful command. @asvetlov @ethanfurman here is my proposed proper fix.

**Comment by asvetlov** [comment]
Let me check on the buildbot fleet. Just in case.

File: Lib/test/test_pty.py [hunk 1]
```diff
@@ -75,20 +75,21 @@ def _readline(fd):
...
-            # Actually the file has complex multi-line structure,
-            # these is no need to parse it for Gentoo check
-            if 'gentoo' in os_release.read_text().lower():
-                # bpo-41818:
-                # Gentoo passes the test,
-                # all other tested Linux distributions fail.
-                # Should not apply @unittest.expectedFailure() on Gentoo
-                # to keep the buildbot fleet happy.
-                return fun
+    # PLATFORM = platform.system()
```
Comment (asvetlov) [inline Lib/test/test_pty.py]: Drop this comment please, it is not required anymore

File: Lib/test/test_pty.py [hunk 2]
```diff
@@ -166,10 +167,12 @@ def test_openpty(self):
         if self.stdin_rows != None and self.stdin_cols != None:
             try:
                 debug("Setting pty.STDIN_FILENO window size")
-                # Set number of columns and rows to be the
-                # floors of 1/5 of respective original values
-                target_stdin_winsz = struct.pack("HHHH", self.stdin_rows//5,
-                                                 self.stdin_cols//5, 0, 0)
+                debug(f"original size: (rows={self.stdin_rows}, cols={self.stdin_cols})")
+                target_stdin_rows = self.stdin_rows + 1
```
Comment (asvetlov) [inline Lib/test/test_pty.py:159]: Please add a few words to describe the resize trick and why it is necessary.

**Review by asvetlov — CHANGES_REQUESTED** [review]
Looks good, a tiny polishing is needed

**Comment by 8vasu** [comment]
I have made the requested changes; please review again. Also, I have rebased so that there is only one clean commit.

File: Lib/test/test_pty.py (hunk 2 above)
Comment (8vasu) [reply Lib/test/test_pty.py:159]: Does this look good?

File: Lib/test/test_pty.py (hunk 1 above)
Comment (8vasu) [reply Lib/test/test_pty.py]: Done.

File: Lib/test/test_pty.py (hunk 2 above)
Comment (8vasu) [reply Lib/test/test_pty.py:159]: I have added some comments. Please let me know if this looks good.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Cool, thanks!

**Comment by 8vasu** [comment]
You're welcome!

**Comment by 8vasu** [comment]
The next pull request will make some additions to the tty module as preparation for making changes to pty.

**Comment by asvetlov** [comment]
@8vasu please also take a look at `test_master_read()` failure on Solaris reported by https://bugs.python.org/issue41818#msg381930

We need to suppress the error for the problematic platform at least if there is no idea why SunOS behaves differently.

**Comment by 8vasu** [comment]
@asvetlov I am working on it now :) I have posted a comment in the bpo issue about the details.
